tutorials/haskell/haskell-time-tutorial/tutorial.md

```diff
@@ -0,0 +1,453 @@
+pattern December = 12
+```
+
+If you’re curious about why the `{-# COMPLETE #-}` pragma is being
+used here’s a [link][complete-pragma] to a complete explanation.
+
+Now if we need to pass a `MonthOfYear` as a parameter to a function
+we could write something like:
+
+```haskell
+foo November or foo 11 --Both will be equivalent
+```
+
+We’ll use this pattern in a function that basically takes a month and
+a day and yields if the corresponding Date is a holiday.
+
+```haskell
+dateToHoliday :: MonthOfYear -> DayOfMonth -> String
+dateToHoliday January  _   = "Happy new year!"
+dateToHoliday November 1   = "Let's eat colada morada"
+dateToHoliday December 25  = "Merry Christmas jojojo"
+dateToHoliday _ _          = "Probably just a regular day"
+```
+
+Now you're probably wondering what is “colada morada”? Haha no worries,
+it’s an ecuatorian tradition in Ecuador and we eat this meal that is
+purple corn based with spices and fruits, pretty amazing to be honest.
+
+Back to Haskell, I think it’s pretty clear what the `dateToHoliday`
+function does, the first clause matches all days of January, second
+one matches the All Saints holiday, the third one matches for Christmas
+and the last one it’s the fallback for all other days. I think using
+the `MonthOfYears` patterns makes it a lot more readable. Having something
+like:
+
+```haskell
+dateToHoliday 12 25  = "Merry Christmas jojojo"
+```
+
+Will not be as straightforward as the function that we have right now.
+
+I haven’t mentioned it yet, but the patterns that we built previously
+are called bidirectional patterns, why? Because we can use them as
+expressions as well, so something like this is totally valid.
+
+``` bash
+ghci> dateToHoliday January 12
+"Happy new year!"
+ghci> dateToHoliday November 1
+"Let's eat colada morada"
+```
+
+
+So now we can use `11`, `November` in the `MonthOfYear` context
+interchangeably across our code.
+
+
+## Stack Builders contribution to Haskell’s time library
+
+Well all of this preamble brings us to the contribution itself.
+Pattern synonyms is what we used when working on `time`’s new features.
+You can see the complete [pull request][PR-1](PR). I have to clarify
+something, the PR has some different code than the one we’re going to
+check here, since the maintainer decided to do some renaming and changed
+the original PR’s code, you can check the changes [here][maintainer-PR].
+Also, she added something extra to the boundaries that we had in mind at
+first and now we can get the days that belong to a period and the period length.
+
+First of all we started defining the new type class `DayPeriod`:
+
+```haskell
+class Ord p => DayPeriod p where
+   -- | Returns the first 'Day' in a period of days.
+   periodFirstDay :: p -> Day
+   -- | Returns the last 'Day' in a period of days.
+   periodLastDay :: p -> Day
+   -- | Get the period this day is in.
+   dayPeriod :: Day -> p
+```
+
```

Comment:
Before the next paragraph, can you add a short explanation about what the type class includes? Is there any law that instances should satisfy?

Reply:
Nice I will add that info.

Reply:
Just one question, by laws you mean class constraints ? the `Ord` one to be more concrete?

Reply:
No, I mean are there properties that instances of `DayPeriod` should satisfy? Here are some examples:

* `for all p. periodFirstDay p <= periodLastDay p`
* `for all p. dayPeriod (periodFirstDay p) == p`
* `for all p. dayPeriod (periodLastDay p) == p`
* `for all d. periodFirstDay (dayPeriod d) <= d`
* `for all d. periodLastDay (dayPeriod d) >= d`

I don't know which ones are required in order to make it a valid instance.

Reply:
@jpvillaisaza good idea, I think this is an excellent opportunity to showcase some of the property tests introduced in the [second PR](https://github.com/haskell/time/pull/182)

Reply:
I didn't mean all of the laws should be added, I'm not sure which ones are required to have a valid instance.

(This is about the things added below, but I'm adding the comment as part of this thread.)